src/signal_handler.c

```diff
@@ -0,0 +1,67 @@
+/**
+ * Copyright (C) 2021 Genome
+ * Research Ltd. All rights reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * @file signal_handler.c
+ * @author Michael Kubiak <mk35@sanger.ac.uk>
+ */
+
+#include "signal_handler.h"
+#include "baton.h"
+
+#include <signal.h>
+
+int signals[] = {SIGINT, SIGQUIT, SIGHUP, SIGTERM, SIGPIPE, 0};
+int exit_flag;
+
+void handle_signal(int signal){
+  switch (signal) {
+  case SIGINT:
+    exit_flag = 2;
+    break;
+  case SIGTERM:
+  case SIGUSR1:
+  case SIGUSR2:
+    exit_flag = 3;
+    break;
+  case SIGPIPE:
+    exit_flag = 4;
+    break;
+  default:
+    exit_flag = 1;
```

Comment:
There are a couple of signals we overlooked where the default behaviour is to stop / continue or to ignore the signal, so I don't think we should add behaviour changing that. i.e. SIGSTP, SIGSTOP / SIGCONT and SIGCHLD.

Reply:
What are you suggesting?  Only signals included in the `signals` array will have their action changed (see L58).  There are a couple of changes that I will make: SIGQUIT should have it's own case and SIGUSR1/2 should be in `signals`.  Is there anything I have missed?

Reply:
Sorry - I'm trying to do too many things at once. Those signals are not in the array, so they won't be caught. Please ignore my comment!

Reply:
And yes, I agree with your other amendments that I didn't spot.